rclcpp/include/rclcpp/executor.hpp

```diff
@@ -76,9 +77,97 @@ class Executor
   virtual void
   spin() = 0;
 
+  /// Add a callback group to an executor.
+  /**
+   * An executor can have zero or more callback groups which provide work during `spin` functions.
+   * When an executor attempts to add a callback group, the executor checks to see if it is already
+   * associated with another executor.
+   * If it is, than an exception is thrown.
+   * Otherwise, the callback group is added to the executor.
+   *
+   * \param[in] group_ptr a shared ptr that points to a callback group
+   * \param[in] node_ptr a shared pointer that points to a node base interface
+   * \param[in] notify True to trigger the interrupt guard condition during this function. If
+   * the executor is blocked at the rmw layer while waiting for work and it is notified that a new
+   * callback group was added, it will wake up.
+   */
+  RCLCPP_PUBLIC
+  virtual void
+  add_callback_group(
+    rclcpp::CallbackGroup::SharedPtr group_ptr,
+    rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr,
+    bool notify = true);
+
+  /// Add vector of callback groups to an executor.
+  /**
+   * An executor can have zero or more callback groups which provide work during `spin` functions.
+   * You can pass a vector of shared pointers that point to callback groups and a shared pointer
+   * to the node base interface that owns the callback groups. Each callback group in the vector
+   * will be passed individually to the `add_callback_group(group_ptr, node_ptr, notify)`.
+   * This is a helper function that allows a user to pass a vector of callback groups and one
+   * node base interface in one call.
+   *
+   * \param[in] group_ptrs a vector of shared ptrs that point to callback groups
+   * \param[in] node_ptr a shared pointer that points to a node base interface
+   * \param[in] notify True to trigger the interrupt guard condition during this function. If
+   * the executor is blocked at the rmw layer while waiting for work and it is notified that a new
+   * callback group was added, it will wake up.
+   */
+  RCLCPP_PUBLIC
+  virtual void
+  add_callback_groups(
```

Comment:
I think we should drop this signature too, since there's no symmetric `remove_callback_groups` which takes a vector and I just don't think it is needed as it is simply a for_each under the hood.